pkg/scannerv4/mappers/mappers.go

```diff
@@ -658,45 +705,69 @@ type severityValues struct {
 	v2Score  float32
 	v3Vector string
 	v3Score  float32
+	source   v4.VulnerabilityReport_Vulnerability_CVSS_Source
+	url      string
 }
 
-// severityAndScores returns the severity and scores information out of a
-// ClairCore vulnerability. The returned information is dependent on the
-// underlying updater.
-func severityAndScores(ctx context.Context, vuln *claircore.Vulnerability, nvdVuln *nvdschema.CVEAPIJSON20CVEItem) (severityValues, error) {
+// severityAndCvssMetrics processes the severity and CVSS scores
+// for a given vulnerability based on the updater type (RHEL, NVD, OSV, etc.).
+// The returned data includes CVSS metrics, ACS preferred severity values, and potential errors
+// encountered during the calculation process for each source.
+func severityAndCvssMetrics(ctx context.Context, vuln *claircore.Vulnerability, nvdVuln *nvdschema.CVEAPIJSON20CVEItem) ([]*v4.VulnerabilityReport_Vulnerability_CVSS, severityValues, error) {
 	ctx = zlog.ContextWithValues(ctx,
-		"component", "mappers/severityAndScores",
+		"component", "mappers/severityAndCvssMetrics",
 		"updater", vuln.Updater,
 		"vuln_id", vuln.ID,
 		"vuln_name", vuln.Name,
 	)
-
+	var cvssMetrics []*v4.VulnerabilityReport_Vulnerability_CVSS
+
+	var sevs []severityValues
+	rhelSev, rhelErr := rhelSeverityAndScores(vuln)
+	if rhelErr == nil {
+		sevs = append(sevs, rhelSev)
+	}
+	nvdSev, nvdErr := nvdSeverityAndScores(vuln, nvdVuln)
+	if nvdErr == nil {
+		sevs = append(sevs, nvdSev)
+	}
+	undecidedSev, undecidedErr := cvssVector(vuln.Severity)
+	if undecidedErr == nil {
+		if strings.HasPrefix(vuln.Updater, osvUpdaterPrefix) {
+			undecidedSev.source = v4.VulnerabilityReport_Vulnerability_CVSS_SOURCE_OSV
+			undecidedSev.url = vuln.Links
+			sevs = append(sevs, undecidedSev)
+		} else if strings.EqualFold(vuln.Updater, constants.ManualUpdaterName) {
+			undecidedSev.source = v4.VulnerabilityReport_Vulnerability_CVSS_SOURCE_UNKNOWN
+			undecidedSev.url = vuln.Links
+			sevs = append(sevs, undecidedSev)
+		}
+	}
```

Comment:
The conditions in this block are also executed below in the switch statement - consider cleaning it up (if it makes sense to do so).

Reply:
Here we only want to update the sources based on updater name before generating the CVSS metrics list but we don't want to return from the method.

Reply:
Actually I've thought of that but we still need the severity from the primary sev after calculate all the the CVSS metrics:
```
vuln := &v4.VulnerabilityReport_Vulnerability{
			Id:                 v.ID,
			Name:               name,
			Description:        description,
			Issued:             issued,
			Link:               v.Links,
			Severity:           primarySev.severity,
			...
			FixedInVersion:     fixedInVersion(v),
		}
```

Reply:
Exactly, and if CVE metrics contains the severity, you would just need to do something like:
```
vuln := &v4.VulnerabilityReport_Vulnerability{
			Id:                 v.ID,
			Name:               name,
			Description:        description,
			Issued:             issued,
			Link:               v.Links,
			Severity:           cvemetrics[primaryIndex].Severity
			...
			FixedInVersion:     fixedInVersion(v),
		}
```

Reply:
I have updated `severityAndCvssMetrics()` to avoid double convert the severity into a cve metric twice. Let me know what you think.